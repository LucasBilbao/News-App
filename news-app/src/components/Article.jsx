import React from 'react';
import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { getTimeFormatted } from '../features/timeFormatted';
import placeholder from '..//assets/img-placeholder.jpg';
import { arrIsEmpty } from '../features/arrayIsEmpty';
import PropTypes from 'prop-types';

function Article({ articles }) {
  let articlesRemoved = 0;

  const { filters } = useSelector(state => state.filters);
  const { title } = useSelector(state => state.title);

  const filterArticleSources = article => {
    if (arrIsEmpty(filters)) {
      return article;
    } else {
      if (filters.includes(article.source.name)) {
        return true;
      } else {
        articlesRemoved += 1;
        return false;
      }
    }
  };

  const filterTitles = article => {
    if (title === '') {
      return article;
    } else {
      if (article.title.toLowerCase().includes(title.toLowerCase())) {
        return true;
      } else {
        articlesRemoved += 1;
        return false;
      }
    }
  };

  // console.log(articles[0]);

  return (
    <div data-test="component-article">
      {articles
        .filter(filterArticleSources)
        .filter(filterTitles)
        .map((article, index) => {
          return (
            <article key={index}>
              <div className="info">
                <p>{article.source.name}</p>
                <p>{getTimeFormatted(new Date(article.publishedAt))}</p>
              </div>
              {article.image ? (
                <img
                  src={article.image}
                  alt="loading..."
                  onError={e => {
                    e.target.src = placeholder;
                  }}
                />
              ) : (
                <img src={placeholder} />
              )}
              <div className="content">
                <h1 data-test="article-title">{article.title}</h1>
                <p data-test="article-description">{article.description}</p>
                <Link to={`/News-App/article/${article.title}`}>Read More</Link>
              </div>
            </article>
          );
        })}
      {articlesRemoved === articles.length && (
        <h1 style={{ color: 'white' }}>There are no articles...</h1>
      )}
    </div>
  );
}

Article.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.object),
};

export default Article;
