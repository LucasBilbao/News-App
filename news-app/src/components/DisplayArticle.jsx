import React from 'react';
import { useParams } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import PulseLoader from 'react-spinners/PulseLoader';
import placeholder from '..//assets/img-placeholder.jpg';

function DisplayArticle({ articles, loading, error }) {
  const { id } = useParams();

  const flexDisplay = {
    display: 'flex',
    width: '100%',
    justifyContent: 'center',
    marginTop: '100px',
  };

  const style = {
    color: 'white',
    width: '100%',
    textAlign: 'center',
  };

  if (loading) {
    return (
      <PulseLoader
        css={flexDisplay}
        color={'#7ebffb'}
        loading={loading}
        size={25}
      />
    );
  } else if (error) {
    return (
      <div className="wrapper">
        <h1 style={style}>{error}</h1>;
      </div>
    );
  }
  return (
    <div>
      <Link to="/News-App/" className="back">
        <button>Back</button>
      </Link>
      <div className="wrapper">
        {articles
          .filter(article => article.title === decodeURI(id))
          .map(article => {
            return (
              <div key={article.url} className="article">
                <h1 data-test="article-title">{article.title}</h1>
                {article.image ? (
                  <img src={article.image} alt="loading..." />
                ) : (
                  <img src={placeholder} />
                )}
                <p>{article.content}</p>
              </div>
            );
          })}
      </div>
    </div>
  );
}

DisplayArticle.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool,
  error: PropTypes.string,
};

export default DisplayArticle;
