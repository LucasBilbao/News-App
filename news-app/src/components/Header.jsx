import React from 'react';

function Header() {
  return (
    <header>
      <h1 data-test="header-title" className="title">
        PewNews
      </h1>
      <h2 data-test="header-subtitle" className="subtitle">
        NEWS
      </h2>
    </header>
  );
}

export default Header;
