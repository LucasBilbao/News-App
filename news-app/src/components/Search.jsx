import React from 'react';
import { useDispatch } from 'react-redux';
import { insertSearch, clearSearch } from '../store/actions/searchTitle';

function Search() {
  const dispatch = useDispatch();
  function handleSubmit(e) {
    e.preventDefault();
    const value = e.target.firstChild.value;
    if (value !== '') {
      dispatch(insertSearch(value));
    } else {
      dispatch(clearSearch());
    }
    e.target.firstChild.value = '';
  }

  return (
    <div className="search">
      <form data-test="search-form" action="submit" onSubmit={handleSubmit}>
        <input
          data-test="form-input"
          type="text"
          placeholder="Title"
          className="search-bar"
        />
        <button data-test="form-button" className="search-btn">
          Search
        </button>
      </form>
    </div>
  );
}

export default Search;
