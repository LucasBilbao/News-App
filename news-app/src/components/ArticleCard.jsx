import React from 'react';
import PropTypes from 'prop-types';
import PulseLoader from 'react-spinners/PulseLoader';
import Article from './Article';
import { arrIsEmpty } from '../features/arrayIsEmpty';
import { useSelector } from 'react-redux';

function ArticleCard({ articles, loading, error }) {
  const flexDisplay = {
    display: 'flex',
    width: '100%',
    justifyContent: 'center',
  };

  const { filters } = useSelector(state => state.filters);
  const { title } = useSelector(state => state.title);

  if (loading) {
    return (
      <PulseLoader
        css={flexDisplay}
        color={'#7ebffb'}
        loading={loading}
        size={25}
      />
    );
  } else if (error) {
    return <h1 data-test="error" style={{ color: 'white' }}>{error}</h1>;
  }
  return (
    <div>
      <p data-test="searched-title" style={{ color: 'white' }}>
        Titles:{' '}
        {title === '' ? (
          <span style={{ color: '#c6ff50' }}>All</span>
        ) : (
          `${title}`
        )}
      </p>
      <p data-test="searched-sources" style={{ color: 'white' }}>
        Sources:{' '}
        {arrIsEmpty(filters) ? (
          <span style={{ color: '#c6ff50' }}>All</span>
        ) : (
          `${filters}`
        )}
      </p>
      <Article articles={articles}></Article>
    </div>
  );
}

ArticleCard.propTypes = {
  articles: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool,
  error: PropTypes.string,
};

export default ArticleCard;
