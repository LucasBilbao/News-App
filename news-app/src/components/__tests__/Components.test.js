/* eslint no-undef: 0 */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import articles from './assets/articleTest.json';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import configureStore from 'redux-mock-store';

// Import Components
import Article from '../Article';
import Search from '../Search';
import DisplayArticle from '../DisplayArticle';
import ArticleCard from '../ArticleCard';
import Header from '../Header';

Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureStore([]);

describe('Test 5 components', () => {
  // Article Tests
  describe('Article Tests', () => {
    describe('1. Test Article with no filter or specified title', () => {
      const props = {
        articles: articles,
      };

      let component;

      const storage = mockStore({
        filters: {
          filters: [],
        },
        title: {
          title: '',
        },
      });

      beforeEach(() => {
        component = mount(
          <Router>
            <Provider store={storage}>
              s<Article {...props}></Article>
            </Provider>
          </Router>
        );
      });

      it('1. Match snapshot with the given articles', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is Truthy', () => {
        expect(component).toBeTruthy();
      });

      it('3. Check the title displayed', () => {
        const articleComponentTitle = component.find(
          '[data-test="article-title"]'
        );
        expect(articleComponentTitle.text()).toBe(articles[0].title);
      });

      it('4. Check the description displayed', () => {
        const articleComponentDescription = component.find(
          '[data-test="article-description"]'
        );
        expect(articleComponentDescription.text()).toBe(
          articles[0].description
        );
      });
    });

    describe('2. Test Articles  with specified filter and title and do not match', () => {
      const props = {
        articles: articles,
      };

      let component;

      const storage = mockStore({
        filters: {
          filters: ['CNN'],
        },
        title: {
          title: 'Covid-19',
        },
      });

      beforeEach(() => {
        component = mount(
          <Router>
            <Provider store={storage}>
              <Article {...props}></Article>
            </Provider>
          </Router>
        );
      });
      it('1. Match snapshot with the given articles', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is Truthy', () => {
        expect(component).toBeTruthy();
      });
    });

    describe('3. Test Articles  with specified filter and title and matches', () => {
      const props = {
        articles: articles,
      };

      let component;

      const storage = mockStore({
        filters: {
          filters: ['The Verge'],
        },
        title: {
          title: articles[0].title,
        },
      });

      beforeEach(() => {
        component = mount(
          <Router>
            <Provider store={storage}>
              <Article {...props}></Article>
            </Provider>
          </Router>
        );
      });
      it('1. Match snapshot with the given articles', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is Truthy', () => {
        expect(component).toBeTruthy();
      });

      it('3. Check the title displayed', () => {
        const articleComponentTitle = component.find(
          '[data-test="article-title"]'
        );
        expect(articleComponentTitle.text()).toBe(articles[0].title);
      });

      it('4. Check the description displayed', () => {
        const articleComponentDescription = component.find(
          '[data-test="article-description"]'
        );
        expect(articleComponentDescription.text()).toBe(
          articles[0].description
        );
      });
    });

    describe('4. Test Articles with specified filter and title, only filter matches', () => {
      const props = {
        articles: articles,
      };

      let component;

      const storage = mockStore({
        filters: {
          filters: ['The Verge'],
        },
        title: {
          title: 'vdsjvvsd',
        },
      });

      beforeEach(() => {
        component = mount(
          <Router>
            <Provider store={storage}>
              <Article {...props}></Article>
            </Provider>
          </Router>
        );
      });
      it('1. Match snapshot with the given articles', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is Truthy', () => {
        expect(component).toBeTruthy();
      });
    });
  });

  // Search Tests
  describe('Search Tests', () => {
    describe('1. Test Search components', () => {
      const storage = mockStore({
        title: {
          title: '',
        },
      });

      let component;

      beforeEach(() => {
        component = mount(
          <Provider store={storage}>
            <Search></Search>
          </Provider>
        );
      });

      it('1. Check Search button content', () => {
        const searchBtn = component.find('[data-test="form-button"]');
        expect(searchBtn.text()).toBe('Search');
      });

      it('2. Check the input is editable', () => {
        const input = component.find('[data-test="form-input"]');
        input.instance().value = 'Changed';
        expect(input.instance().value).toBe('Changed');
      });

      it('2. Check submission', () => {
        const input = component.find('[data-test="form-input"]');
        input.instance().value = 'Changed';

        const searchBtn = component.find('[data-test="search-form"]');
        searchBtn.simulate('submit');

        expect(input.instance().value).toBe('');
      });

      it('3. Check that form submits without changing empty input', () => {
        const input = component.find('[data-test="form-input"]');

        const searchBtn = component.find('[data-test="search-form"]');
        searchBtn.simulate('submit');

        expect(input.instance().value).toBe('');
      });
    });
  });

  // DisplayArticle Tests
  describe('DisplayArticle Tests', () => {
    describe('1. Check loading', () => {
      const props = {
        articles: [],
        loading: true,
        error: null,
      };

      let component;

      beforeEach(() => {
        component = mount(
          <Router>
            <DisplayArticle {...props}></DisplayArticle>
          </Router>
        );
      });

      it('1. Match snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });
    });

    describe('2. Check error', () => {
      const props = {
        articles: [],
        loading: false,
        error: 'An Error has ocurred',
      };

      let component;

      beforeEach(() => {
        component = mount(
          <Router>
            <DisplayArticle {...props}></DisplayArticle>
          </Router>
        );
      });

      it('1. Match snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });
    });

    describe('2. Test Displayed title and content of the Article ', () => {
      const props = {
        articles: articles,
        loading: false,
        error: null,
      };

      let component;

      const id = encodeURI(articles[0].title);

      beforeEach(() => {
        component = mount(
          <Router>
            <DisplayArticle
              {...props}
              match={{ params: { id: id } }}
            ></DisplayArticle>
          </Router>
        );
      });

      it('1. Match snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });
    });
  });

  // Article Card Tests
  describe('Article Card Tests', () => {
    describe('1. Check loading', () => {
      const props = {
        articles: articles,
        loading: false,
        error: null,
      };

      const storage = mockStore({
        filters: {
          filters: [],
        },
        title: {
          title: '',
        },
      });

      let component;

      beforeEach(() => {
        component = mount(
          <Provider store={storage}>
            <Router>
              <ArticleCard {...props}></ArticleCard>
            </Router>
          </Provider>
        );
      });

      it('1. Match Snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });
    });

    describe('2. Check loading', () => {
      const props = {
        articles: articles,
        loading: true,
        error: null,
      };

      const storage = mockStore({
        filters: {
          filters: [],
        },
        title: {
          title: '',
        },
      });

      let component;

      beforeEach(() => {
        component = mount(
          <Provider store={storage}>
            <Router>
              <ArticleCard {...props}></ArticleCard>
            </Router>
          </Provider>
        );
      });

      it('1. Match Snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });
    });

    describe('3. Check error', () => {
      const props = {
        articles: articles,
        loading: false,
        error: 'An Error has ocurred',
      };

      const storage = mockStore({
        filters: {
          filters: [],
        },
        title: {
          title: '',
        },
      });

      let component;

      beforeEach(() => {
        component = mount(
          <Provider store={storage}>
            <Router>
              <ArticleCard {...props}></ArticleCard>
            </Router>
          </Provider>
        );
      });
      it('1. Match Snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });

      it('3. check error message', () => {
        const errMsg = component.find('[data-test="error"]');
        expect(errMsg.text()).toBe(props.error);
      });
    });

    describe('4. Check when filter and title are specified', () => {
      const props = {
        articles: articles,
        loading: false,
        error: null,
      };

      const storage = mockStore({
        filters: {
          filters: ['CNN', 'BBC News'],
        },
        title: {
          title: 'COVID-19',
        },
      });

      let component;

      beforeEach(() => {
        component = mount(
          <Provider store={storage}>
            <Router>
              <ArticleCard {...props}></ArticleCard>
            </Router>
          </Provider>
        );
      });

      it('1. Match Snapshot', () => {
        expect(component).toMatchSnapshot();
      });

      it('2. Check if component is truthy', () => {
        expect(component).toBeTruthy();
      });

      it('3. Check searched titles displayed', () => {
        const titleSearched = component.find('[data-test="searched-title"]');
        const wantedOutcome = `Titles: ${storage.getState().title.title}`;
        expect(titleSearched.text()).toBe(wantedOutcome);
      });

      it('4. Check searched sources displayed', () => {
        const sourcesSearched = component.find(
          '[data-test="searched-sources"]'
        );
        const wantedOutcome = `Sources: ${storage
          .getState()
          .filters.filters.join(',')}`;
        expect(sourcesSearched.text()).toBe(wantedOutcome);
      });
    });
  });

  // Header Tests
  describe('Header Tests', () => {
    describe('1. Test Header title and subtitle', () => {
      let component;

      beforeEach(() => {
        component = mount(<Header></Header>);
      });

      it('1. Test title', () => {
        const title = component.find('[data-test="header-title"]');
        expect(title.text()).toBe('PewNews');
      });

      it('1. Test subtitle', () => {
        const title = component.find('[data-test="header-subtitle"]');
        expect(title.text()).toBe('NEWS');
      });
    });
  });
});
