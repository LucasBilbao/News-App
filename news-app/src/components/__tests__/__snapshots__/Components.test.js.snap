// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test 5 components Article Card Tests 1. Check loading 1. Match Snapshot 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Card Tests 2. Check loading 1. Match Snapshot 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Card Tests 3. Check error 1. Match Snapshot 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Card Tests 4. Check when filter and title are specified 1. Match Snapshot 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Tests 1. Test Article with no filter or specified title 1. Match snapshot with the given articles 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Tests 2. Test Articles  with specified filter and title and do not match 1. Match snapshot with the given articles 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Tests 3. Test Articles  with specified filter and title and matches 1. Match snapshot with the given articles 1`] = `ReactWrapper {}`;

exports[`Test 5 components Article Tests 4. Test Articles with specified filter and title, only filter matches 1. Match snapshot with the given articles 1`] = `ReactWrapper {}`;

exports[`Test 5 components DisplayArticle Tests 1. Check loading 1. Match snapshot 1`] = `ReactWrapper {}`;

exports[`Test 5 components DisplayArticle Tests 2. Check error 1. Match snapshot 1`] = `ReactWrapper {}`;

exports[`Test 5 components DisplayArticle Tests 2. Test Displayed title and content of the Article  1. Match snapshot 1`] = `ReactWrapper {}`;
